x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/transport/TransportFindStructureAction.java

```diff
@@ -20,27 +21,39 @@
 import org.elasticsearch.xpack.textstructure.structurefinder.TextStructureOverrides;
 
 import java.io.InputStream;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+
+import static org.elasticsearch.common.util.concurrent.EsExecutors.DIRECT_EXECUTOR_SERVICE;
+import static org.elasticsearch.threadpool.ThreadPool.Names.GENERIC;
 
 public class TransportFindStructureAction extends HandledTransportAction<FindStructureAction.Request, FindStructureResponse> {
 
     private final ThreadPool threadPool;
+    private final ExecutorService executorService;
 
     @Inject
     public TransportFindStructureAction(TransportService transportService, ActionFilters actionFilters, ThreadPool threadPool) {
-        super(FindStructureAction.NAME, transportService, actionFilters, FindStructureAction.Request::new, threadPool.generic());
+        super(FindStructureAction.NAME, transportService, actionFilters, FindStructureAction.Request::new, DIRECT_EXECUTOR_SERVICE);
         this.threadPool = threadPool;
+        this.executorService = threadPool.generic();
     }
 
     @Override
     protected void doExecute(Task task, FindStructureAction.Request request, ActionListener<FindStructureResponse> listener) {
+        // workaround for https://github.com/elastic/elasticsearch/issues/97916 - TODO remove this when we can
+        // when the workaround is removed, change the value that this class's constructor passes to the super constructor from
+        // DIRECT_EXECUTOR_SERVICE back to threadpool.generic() so that we continue to fork off of the transport thread.
+        var buildTextStructureResponse = ActionRunnable.supply(listener, () -> buildTextStructureResponse(request));
         try {
-            listener.onResponse(buildTextStructureResponse(request));
-        } catch (Exception e) {
-            listener.onFailure(e);
+            executorService.execute(buildTextStructureResponse);
```

Comment:
No need for the try/catch, `executorService.execute` won't throw anything if you give it an `AbstractRunnable`. And anyway this whole thing is running in a try/catch further up the stack.